File: packages/python/halt/algorithms/leaky_bucket.py
"""Leaky bucket rate limiting algorithm."""

import time
from typing import Optional
from halt.core.decision import Decision


class LeakyBucket:
    """Leaky bucket algorithm for rate limiting.
    
    The leaky bucket algorithm processes requests at a constant rate,
    smoothing out bursts. Requests are added to a bucket and leak out
    at a constant rate. Excellent for traffic shaping.
    """
    
    def __init__(self, capacity: int, leak_rate: float, window: int) -> None:
        """Initialize leaky bucket.
        
        Args:
            capacity: Maximum bucket capacity (max burst size)
            leak_rate: Rate at which requests leak (requests per second)
            window: Time window for rate calculation (seconds)
        """
        self.capacity = capacity
        self.leak_rate = leak_rate  # requests per second
        self.window = window
    
    def check_and_consume(
        self,
        current_level: float,
        last_leak: float,
        cost: int,
        now: Optional[float] = None,
    ) -> tuple[Decision, float, float]:
        """Check if request is allowed and update bucket level.
        
        Args:
            current_level: Current water level in bucket
            last_leak: Timestamp of last leak
            cost: Number of requests to add
            now: Current timestamp (defaults to time.time())
        
        Returns:
            Tuple of (Decision, new_level, new_last_leak)
        """
        if now is None:
            now = time.time()
        
        # Calculate how much has leaked since last check
        elapsed = now - last_leak
        leaked = elapsed * self.leak_rate
        
        # Update bucket level (can't go below 0)
        new_level = max(0, current_level - leaked)
        
        # Calculate when bucket will be empty
        if new_level > 0:
            time_to_empty = new_level / self.leak_rate
            reset_at = int(now + time_to_empty)
        else:
            reset_at = int(now)
        
        # Check if we have capacity
        if new_level + cost <= self.capacity:
            # Allow request - add to bucket
            new_level += cost
            remaining = int(self.capacity - new_level)
            
            return (
                Decision(
                    allowed=True,
                    limit=self.capacity,
                    remaining=remaining,
                    reset_at=reset_at,
                ),
                new_level,
                now,
            )
        else:
            # Block request - bucket is full
            # Calculate when there will be enough capacity
            space_needed = (new_level + cost) - self.capacity
            retry_after = int(space_needed / self.leak_rate) + 1
            
            return (
                Decision(
                    allowed=False,
                    limit=self.capacity,
                    remaining=0,
                    reset_at=reset_at,
                    retry_after=retry_after,
                ),
                new_level,
                now,
            )
    
    def initial_state(self, now: Optional[float] = None) -> tuple[float, float]:
        """Get initial state for a new key.
        
        Args:
            now: Current timestamp (defaults to time.time())
        
        Returns:
            Tuple of (level, last_leak)
        """
        if now is None:
            now = time.time()
        return (0.0, now)

File: packages/typescript/src/algorithms/leaky-bucket.ts
/**
 * Leaky bucket rate limiting algorithm.
 */

import { Decision } from '../core/decision';

export interface LeakyBucketState {
    level: number;
    lastLeak: number;
}

export class LeakyBucket {
    private capacity: number;
    private leakRate: number;
    private window: number;

    constructor(capacity: number, leakRate: number, window: number) {
        this.capacity = capacity;
        this.leakRate = leakRate; // requests per second
        this.window = window;
    }

    /**
     * Check if request is allowed and update bucket level.
     */
    checkAndConsume(
        currentLevel: number,
        lastLeak: number,
        cost: number,
        now: number = Date.now() / 1000
    ): { decision: Decision; newLevel: number; newLastLeak: number } {
        // Calculate how much has leaked since last check
        const elapsed = now - lastLeak;
        const leaked = elapsed * this.leakRate;

        // Update bucket level (can't go below 0)
        let newLevel = Math.max(0, currentLevel - leaked);

        // Calculate when bucket will be empty
        let resetAt: number;
        if (newLevel > 0) {
            const timeToEmpty = newLevel / this.leakRate;
            resetAt = Math.floor(now + timeToEmpty);
        } else {
            resetAt = Math.floor(now);
        }

        // Check if we have capacity
        if (newLevel + cost <= this.capacity) {
            // Allow request - add to bucket
            newLevel += cost;
            const remaining = Math.floor(this.capacity - newLevel);

            return {
                decision: {
                    allowed: true,
                    limit: this.capacity,
                    remaining,
                    resetAt,
                },
                newLevel,
                newLastLeak: now,
            };
        } else {
            // Block request - bucket is full
            // Calculate when there will be enough capacity
            const spaceNeeded = newLevel + cost - this.capacity;
            const retryAfter = Math.floor(spaceNeeded / this.leakRate) + 1;

            return {
                decision: {
                    allowed: false,
                    limit: this.capacity,
                    remaining: 0,
                    resetAt,
                    retryAfter,
                },
                newLevel,
                newLastLeak: now,
            };
        }
    }

    /**
     * Get initial state for a new key.
     */
    initialState(now: number = Date.now() / 1000): LeakyBucketState {
        return {
            level: 0,
            lastLeak: now,
        };
    }
}
